src/main/c/ssl.c

```diff
@@ -231,26 +231,35 @@ static const jint supported_ssl_opts = 0
 
 static int ssl_tmp_key_init_rsa(int bits, int idx)
 {
-#if (OPENSSL_VERSION_NUMBER < 0x10100000L) || defined(OPENSSL_USE_DEPRECATED)
-    if (!(SSL_temp_keys[idx] =
-          RSA_generate_key(bits, RSA_F4, NULL, NULL))) {
-#ifdef OPENSSL_FIPS
-        /**
-         * With FIPS mode short RSA keys cannot be
-         * generated.
-         */
-        if (bits < 1024)
-            return 0;
-        else
-#endif
-        return 1;
-    }
-    else {
+#if defined(OPENSSL_FIPS)
+    /**
+     * Short RSA keys cannot be generated in FIPS mode.
+     */
+    if (bits < 1024) {
+        SSL_temp_keys[idx] = NULL;
         return 0;
     }
-#else
-    return 0;
 #endif
+
+    BIGNUM *e = BN_new();
+    RSA *rsa = RSA_new();
+    int ret = 1;
+
+    if (e == NULL ||
+        rsa == NULL ||
+        !BN_set_word(e, RSA_F4) ||
```

Comment:
OpenSSL isn't consistent about return values and sometimes returns negative values too.

I would generally use the boolean-style everywhere except that RSA_generate_key[_ex] is a function that can be redirected to a different method, so I fear that such code might end up returning negative numbers even though the default code does not. It would also be a little hard to criticise such code since the return value for _ex is [not documented](https://www.openssl.org/docs/manmaster/crypto/RSA_generate_key.html).